src/irmin-pack/inode.ml

```diff
@@ -759,6 +759,117 @@ struct
+          | `Node h -> List [ Atom "N"; sexp_of_hash h ]
+
+        let sexp_of_step s =
+          Sexplib.Sexp.Atom (Fmt.str "%a" (Irmin.Type.pp step_t) s)
+
+        let rec list_of_entries a =
+          Array.fold_left
+            (fun acc e ->
+              match e with
+              | None -> acc
+              | Some p -> (
+                  match p.I.target with
+                  | None -> acc
+                  | Some t -> List.rev_append (list_of_t t) acc))
+            [] a
+
+        and list_of_tree t = list_of_entries t.I.entries
+
+        and list_of_stepmap vs =
+          let open Sexplib.Sexp in
+          StepMap.fold
+            (fun s k acc -> List [ sexp_of_step s; sexp_of_kind k ] :: acc)
+            vs []
+
+        and list_of_v v =
+          match v with
+          | I.Values vs -> list_of_stepmap vs
+          | Tree t -> list_of_tree t
+
+        and list_of_t t = list_of_v t.I.v
+
+        let sexp_of_t t =
+          Sexplib.Sexp.(
+            List
+              ([
+                 Atom (string_of_int Conf.entries);
+                 Atom (string_of_int Conf.stable_hash);
+                 sexp_of_lazy_hash t.v.I.hash;
+               ]
+              @ list_of_t t.v))
+
+        exception Misconstructed_Sexp of string
+        exception Wrong_Config of (int * int) * (int * int)
+
+        let ( ?>= ) v = match v with Ok v -> v | Error (`Msg e) -> failwith e
+
+        let step_value_list_of_sexp (s : sexp list) =
+          let rec aux acc = function
+            | [] -> acc
+            | Sexplib.Sexp.(List [ Atom step; List value ]) :: tl ->
+                let value =
+                  match value with
+                  | [ Atom "B"; Atom hash; Atom metadata ] ->
+                      let hash = ?>=(Irmin.Type.of_string hash_t hash) in
+                      let meta =
+                        ?>=(Irmin.Type.of_string metadata_t metadata)
+                      in
+                      `Contents (hash, meta)
+                  | [ Atom "N"; Atom hash ] ->
+                      let hash = ?>=(Irmin.Type.of_string hash_t hash) in
+                      `Node hash
+                  | _ -> raise Exit
+                in
+                aux ((?>=(Irmin.Type.of_string step_t step), value) :: acc) tl
+            | _ -> raise Exit
+          in
+          aux [] s
+
+        let t_of_sexp ts =
+          let open Sexplib.Sexp in
+          match ts with
+          | List
+              (Atom entries :: Atom stable_hash :: Atom hash :: step_value_list)
+            ->
+              let entries = int_of_string entries in
+              let stable_hash = int_of_string stable_hash in
+              if Conf.stable_hash <> stable_hash || Conf.entries <> entries then
+                raise
+                  (Wrong_Config
+                     ((Conf.entries, Conf.stable_hash), (entries, stable_hash)))
+              else (v @@ step_value_list_of_sexp step_value_list, hash)
```

Comment:
Instead of returning `hash` as a string, could you transform it back to a `H.t`?

Reply:
Yep, it's the plan, I just needed it as a string to quickcheck it ;-)